File: Kafka-Java/src/main/java/part1_KafkaJavaProgramming/ProducerWithCallback.java
package part1_KafkaJavaProgramming;

import org.apache.kafka.clients.producer.*;
import org.apache.kafka.common.serialization.StringSerializer;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.Properties;

public class ProducerWithCallback {
    public static void main(String[] args) {
        //Init value and variable
        final Logger logger = LoggerFactory.getLogger(ProducerWithCallback.class);
        String bootstrapServers = "127.0.0.1:9092";

        //Create Producer properties
        Properties properties = new Properties();

        properties.setProperty(ProducerConfig.BOOTSTRAP_SERVERS_CONFIG, bootstrapServers);
        properties.setProperty(ProducerConfig.KEY_SERIALIZER_CLASS_CONFIG, StringSerializer.class.getName());
        properties.setProperty(ProducerConfig.VALUE_SERIALIZER_CLASS_CONFIG, StringSerializer.class.getName());

        //Create the producer & producer record
        KafkaProducer<String, String> producer = new KafkaProducer<String, String>(properties);

        for(int i =0; i < 10; i++){
            ProducerRecord<String, String> producerRecord = new ProducerRecord<String, String>("quickstart-events", "Hello Kafka from Java " + Integer.toString(i));

            //Send data
            producer.send(producerRecord, new Callback() {
                public void onCompletion(RecordMetadata recordMetadata, Exception e) {
                    //executed every time a record is successfully sent or an exception thrown
                    if (e == null){
                        //the record was successfully sent
                        logger.info("Received new data: \n" +
                                "Topic: " + recordMetadata.topic() + "\n" +
                                "Partition: " + recordMetadata.partition() + "\n" +
                                "Offset: " + recordMetadata.offset() + "\n" +
                                "Timestamp: " + recordMetadata.timestamp() + "\n" );
                    }else {
                        logger.error("Error when receiving data", e);
                    }
                }
            });
        }

        producer.flush();
        producer.close();
    }
}

File: Kafka-Scala/src/main/scala/part1_KafkaScalaProgramming/ProducerWithCallBack.scala
package part1_KafkaScalaProgramming

import java.util.Properties

import org.apache.kafka.clients.producer.{Callback, KafkaProducer, ProducerRecord, RecordMetadata}
import org.apache.kafka.common.serialization.StringSerializer
import org.slf4j.LoggerFactory

object ProducerWithCallBack extends App{
  val bootstrapServer = "127.0.0.1:9092"
  val logger = LoggerFactory.getLogger(IntroProducer.getClass)
  //Create properties
  val kafkaProducerProps: Properties = {
    val props = new Properties()
    props.put("bootstrap.servers", bootstrapServer)
    props.put("key.serializer", classOf[StringSerializer].getName)
    props.put("value.serializer", classOf[StringSerializer].getName)
    props
  }
  val producer = new KafkaProducer[String, String](kafkaProducerProps)

  for(index <- 1 to 10){
    //Create producer and record
    val record = new ProducerRecord[String, String]("quickstart-events", s"Hello Kafka from Scala: $index")
    producer.send(record, new Callback {
      override def onCompletion(metadata: RecordMetadata, exception: Exception): Unit = {
        if( exception == null){ // TODO: not scala style, need change !!!
          //the record was successfully sent
          logger.info("Received new data: \n" +
            "Topic: " + metadata.topic() + "\n" +
            "Partition: " + metadata.partition() + "\n" +
            "Offset: " + metadata.offset() + "\n" +
            "Timestamp: " + metadata.timestamp() + "\n" );
        }else {
          logger.error("Error when receiving data", exception);
        }
      }
    })

  }

  producer.close()
}
